CONTRIBUTING.adoc

```diff
@@ -54,194 +66,225 @@ However, to avoid doing too much work all the time, we have checked the generate
-(see Hercules).
+Also, include any relevant meta-information, such as ticket numbers.
+If a commit completely addresses a ticket, you can put that in the headline if you want, but it's fine to just put it in the body.
 
-Hydra will not report a failed status if `release.nix` has errors, we have a Builkite job to check this.
+There is plenty to say on this topic, but broadly the guidelines in link:https://chris.beams.io/posts/git-commit/[this post] are good.
+
+== Commit signing
+
+Set it up if you can, it's relatively easy to do.
+
+== Rebasing, force-pushing, and history
+
+Until a PR is merged, the branch is yours to do with as you will.
+In particular, rebasing and force-pushing is fine.
+Indeed, if you need to update your branch with changes from master, rebasing is typically better than merging.
+
+So please do use this ability where it helps, for example:
+
+* Add low-effort or WIP commits to fix review comments, and then squash them away before merging the PR.
+* If you have already had a PR review, don't rebase away the old commits until the PR is ready to merge, so that the reviewer only has to look at the "new" commits.
+* Rewrite the commits to make the story clearer where possible.
+
+Don't be obsessive about history though: a little bit of effort making the history clear is nice, but you can rapidly hit diminishing returns.
+Use your judgement, but probably don't merge a PR that has commits called "WIP" or "fix"!
+
+If a PR is just a total mess, consider using Github's squash-merge feature.
+
+== Pull requests
+
+All changes go through pull requests, there is no way to push to master.
+
+=== Code review and merging
+
+All pull-requests should be approved by at least one other person.
+We don't enforce this, though: a PR fixing a typo is fine to self-merge, beyond that use your judgement.
+
+As an author, code review is an opportunity for you to get feedback from clear eyes.
+As a reviewer, code review is an opportunity for you to help your colleagues and learn about what they are doing.
+Make the best use of it you can!
+
+* Pick the right reviewer(s). If you don't know who to pick, ask!
+* Respect your reviewers' time. Their time is as valuable as yours, and it's typically more efficient for you
+to spend time explaining or clarifying something in advance than for them to puzzle it out or pose a question.
+* Review code as quickly as you can. Waiting for review is often a blocker for other people, so prioritize it.
+* If you don't understand something, ask. You are as clever as any person who will read this in the future, if it confuses you it's confusing.
+* If someone had to ask about your code, it wasn't clear enough so change it or add a comment.
+* More reviewing is usually helpful. If you think a PR is interesting, you can review it even if nobody asked you to, you will probably have things to contribute and you'll learn something.
```

Comment:
I would add that IME it makes a big difference (especially long-term) if you try to understand what the code is actually doing, rather than just "linting" it. Additionally I think it should be mentioned somewhere that lots of small PRs is much better than one big one as it makes reviewing much easier.